eth_account/_utils/encode_typed_data/encoding_and_hashing.py

```diff
@@ -0,0 +1,221 @@
+            return (type_, to_int(hexstr=value))
+        else:
+            return (type_, to_int(text=value))
+
+    return (type_, value)
+
+
+def find_type_dependencies(type_, types, results=None):
+    if results is None:
+        results = set()
+
+    # a type must be a string
+    if not isinstance(type_, str):
+        raise ValueError(
+            "Invalid find_type_dependencies input: expected string, got "
+            f"{type_} of type {type(type_)}"
+        )
+    # get core type if it's an array type
+    type_ = parse_core_array_type(type_)
+
+    # don't look for dependencies of solidity types
+    if type_ in EIP712_SOLIDITY_TYPES:
+        return results
+
+    # found a type that isn't defined
+    if type_ not in types:
+        raise ValueError(f"No definition of type {type_}")
+
+    # found a type that's already been added
+    if type_ in results:
+        return results
+
+    results.add(type_)
+
+    for field in types[type_]:
+        find_type_dependencies(field["type"], types, results)
+    return results
+
+
+def encode_type(type_: str, types: Dict[str, List[Dict[str, str]]]) -> str:
+    result = ""
+    unsorted_deps = find_type_dependencies(type_, types)
+    if type_ in unsorted_deps:
+        unsorted_deps.remove(type_)
+
+    deps = [type_] + sorted(list(unsorted_deps))
+    for type_ in deps:
+        children_list = []
+        for child in types[type_]:
+            child_type = child["type"]
+            child_name = child["name"]
+            children_list.append(f"{child_type} {child_name}")
+
+        result += f"{type_}({','.join(children_list)})"
+
+    return result
+
+
+def hash_type(type_: str, types: Dict[str, List[Dict[str, str]]]) -> bytes:
+    return keccak(text=encode_type(type_, types))
+
+
+def encode_data(
+    type_: str,
+    types: Dict[str, List[Dict[str, str]]],
+    data: Dict[str, Any],
+) -> bytes:
+    encoded_types: List[str] = ["bytes32"]
+    encoded_values: List[Union[bytes, int]] = [hash_type(type_, types)]
+
+    for field in types[type_]:
+        type, value = encode_field(
+            types,
+            field["name"],
+            field["type"],
+            data.get(field["name"]),
+        )
+        encoded_types.append(type)
+        encoded_values.append(value)
+
+    return encode(encoded_types, encoded_values)
+
+
+def hash_struct(
+    type_: str,
+    types: Dict[str, List[Dict[str, str]]],
+    data: Dict[str, Any],
+) -> bytes:
+    encoded = encode_data(type_, types, data)
+    return keccak(encoded)
+
+
+def hash_EIP712_message(
+    # returns the same hash as `hash_struct`, but automatically determines primary type
+    message_types: Dict[str, List[Dict[str, str]]],
+    message_data: Dict[str, Any],
+) -> bytes:
+    primary_type = get_primary_type(message_types)
+    return keccak(encode_data(primary_type, message_types, message_data))
+
+
+def hash_domain(domain_data: Dict[str, Any]) -> bytes:
+    EIP712_domain_map = {
```

Comment:
nit: [PEP8: N806](https://pypi.org/project/pep8-naming/) wants this function variable to be lowercase, getting a PyCharm warning as well.